libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,185 @@
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+    onFreeDrive();
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, null);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
+    removeLocationUpdates(callback);
+    future.cancel(false);
+    future = null;
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(Looper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        if (rawLocation != null) {
+          enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+            @Override
+            public void run() {
+              callback.onSuccess(LocationEngineResult.create(enhancedLocation));
+            }
+          });
+        }
+      }
+    }, 1500, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  void onNavigationKilled() {
+    if (future != null) {
+      removeLocationUpdates(callback);
+      future.cancel(false);
+      future = null;
+    }
+  }
+
+  // TODO Do we want to call configure as part of EnhancedLocationEngineDecorator construction?
+  void configure(@NonNull File path, @NonNull String version) {
+    OfflineNavigator offlineNavigator =
+      new OfflineNavigator(mapboxNavigator.getNavigator());
+    String tilePath = new File(path, version).getAbsolutePath();
+    offlineNavigator.configure(tilePath, new OnOfflineTilesConfiguredCallback() {
+      @Override
+      public void onConfigured(int numberOfTiles) {
+        Timber.d("DEBUG: onConfigured %d", numberOfTiles);
+        locationEngine.requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
```

Comment:
aren't we starting location engine twice here and in constructor?

Reply:
Good catch! Fixing.